Forta-Bot-Challenge-4/COMP01/src/thresholdCache/thresholdCache.ts

```diff
@@ -0,0 +1,108 @@
+import { LRUCache } from "lru-cache";
+import { ethers } from "ethers";
+
+import { BigNumber } from "@ethersproject/bignumber";
+
+export let THRESHOLD: any = ethers.BigNumber.from("10");
+
+export const amountCache: LRUCache<string, { timestamp: number; amount: any }> = new LRUCache({
+  max: 1000,
+});
+
+export const amountOverThreshold = async (userAddress: string, amount: any, timestamp: number): Promise<number> => {
+  const bigAmount = BigNumber.from(amount);
+  const bigThreshold = BigNumber.from(THRESHOLD);
+
+  // Check if there is an entry in the cache for the current userAddress
+  if (amountCache.has(userAddress)) {
+    const cachedData = amountCache.get(userAddress) || { timestamp: 0, amount: "0" };
+    const totalAmountWithinDay = BigNumber.from(cachedData.amount).add(bigAmount);
+
+    // Update the cache with the provided timestamp and accumulated amount
+    amountCache.set(userAddress, {
+      timestamp: timestamp,
+      amount: totalAmountWithinDay.toString(),
+    });
+
+    // Check if the total amount within the last day is over the threshold
+    if (totalAmountWithinDay.gt(bigThreshold)) {
+      return totalAmountWithinDay.sub(bigThreshold).toNumber();
+    }
+  } else {
+    // If there is no entry in the cache, update the cache with the provided timestamp and amount
+    amountCache.set(userAddress, {
+      timestamp: timestamp,
+      amount: bigAmount.toString(),
+    });
+
+    // Check if the current amount is greater than the threshold
+    if (bigAmount.gt(bigThreshold)) {
+      return bigAmount.sub(bigThreshold).toNumber();
+    }
+  }
+
+  // Return 0 since the total amount within the last day is not over the threshold
+  return 0;
+};
+
+// Function to update the threshold as 20% above the average of the amounts stored in the cache
+export const calculateNewThreshold = (): void => {
```

Comment:
I don't see this function being called in the bot logic, so I'm not sure how you fully intend to implement it. However, a simple way you could have the average would be have global variables to which you just add onto whenever a `supply` is caught.

You could have a global variable named something like `totalSupplySum` to which you just add the `amount` from the event. Then a second global variable named something like `supplyCounter` that you increment by 1 (e.g. `++`) for each `supply` event. Then to find the average, divide `totalSupplySum / supplyCounter`.

This could allow you to circumvent the need to process the values in your cache. And since you would be altering the values of your cache based on your user activity, it could lead to incorrect values.

Reply:
I ended up just removing this function `calculateNewThreshold` and implemented the logic above in the `amountOverThreshold` function in order to make the threshold dynamic.